ibtk/src/utilities/snapshot_utilities.cpp

```diff
@@ -0,0 +1,132 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2022 - 2022 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include <ibtk/snapshot_utilities.h>
+
+#include <HierarchyDataOpsManager.h>
+
+#include <ibtk/namespaces.h> // IWYU pragma: keep
+
+/////////////////////////////// NAMESPACE ////////////////////////////////////
+namespace IBTK
+{
+template <class VariableType>
+void
+fillSnapshotAtTime(const int u_idx,
+                   const double time,
+                   const int scr_idx,
+                   SnapshotCache<VariableType>& snapshot_cache,
+                   Pointer<PatchHierarchy<NDIM> > hierarchy,
+                   const std::string& refine_type,
+                   Pointer<HierarchyDataOpsReal<NDIM, double> > hier_data_ops,
+                   const double period)
+{
+    const std::set<double>& snapshot_times = snapshot_cache.getSnapshotTimes();
+    // If there's only one snapshot, return it
+    if (snapshot_times.size() == 1)
+    {
+        TBOX_ASSERT((*snapshot_times.begin()) == time);
+        snapshot_cache.getSnapshot(u_idx, time, hierarchy, "LINEAR_REFINE", 1.0e-8);
+        return;
+    }
+
+    // Determine the correct snapshot index
+    double snapshot_time_low, snapshot_time_up;
+    double t_low, t_up;
+    auto it_up = snapshot_times.upper_bound(time);
+    if (period == period && it_up == snapshot_times.end())
+    {
+        // upper point is actually the first point
+        snapshot_time_up = (*snapshot_times.begin());
+        t_up = snapshot_time_up + period;
+        snapshot_time_low = (*snapshot_times.end());
+        t_low = snapshot_time_low;
+    }
+    else if (it_up == snapshot_times.end())
+    {
+        TBOX_ERROR("Could not find time points around time " << time << " in snapshot cache!\n");
+    }
+    else
+    {
+        // everything is normal
+        snapshot_time_up = (*it_up);
+        t_up = snapshot_time_up;
+        snapshot_time_low = *(std::next(it_up, -1));
+        t_low = snapshot_time_low;
+    }
+    // Create a valid hier_data_ops object
+    if (!hier_data_ops)
+    {
+        auto hier_math_ops = HierarchyDataOpsManager<NDIM>::getManager();
+        auto var_db = VariableDatabase<NDIM>::getDatabase();
+        Pointer<Variable<NDIM> > var;
+        var_db->mapIndexToVariable(u_idx, var);
+        Pointer<HierarchyDataOpsReal<NDIM, double> > hier_data_ops = hier_math_ops->getOperationsDouble(var, hierarchy);
```

Comment:
```suggestion
        Pointer<HierarchyDataOpsReal<NDIM, double> > hier_data_ops = hier_math_ops->getOperationsDouble(var, hierarchy, true);
```

Reply:
Should we make our own function to get these things that always generates unique objects?